zcash_test_vectors/transaction_zsa.py

```diff
@@ -0,0 +1,216 @@
+        fvk_r = FullViewingKey.from_spending_key(SpendingKey(rand.b(32)))
+        self.recipient = fvk_r.default_d() + bytes(fvk_r.default_pkd())
+        self.value = rand.u64()
+        self.assetBase = zsa_value_base(asset_digest(encode_asset_id(ik, asset_desc)))
+        self.rho = Point.rand(rand).extract()
+        self.rseed = rand.b(32)
+
+    def __bytes__(self):
+        ret = b''
+        ret += bytes(self.recipient)
+        ret += struct.pack('<Q', self.value)
+        ret += bytes(self.assetBase)
+        ret += bytes(self.rho)
+        ret += self.rseed
+
+        return ret
+
+
+class TransactionZSA(TransactionBase):
+    def __init__(self, rand, consensus_branch_id, have_orchard_zsa = True, have_burn = True, have_issuance = True):
+
+        # Since burn is part of the OrchardZSA bundle, ensure that there are no burn fields
+        # when there are no OrchardZSA fields
+        assert have_orchard_zsa or not have_burn
+
+        # All the Transparent and Sapling Transaction Fields are initialized in the super (TransactionBase) class.
+        super().__init__(rand)
+
+        # Common Transaction Fields
+        self.nVersionGroupId = NU7_VERSION_GROUP_ID
+        self.nConsensusBranchId = consensus_branch_id
+
+        # Orchard-ZSA Transaction Fields
+        self.vActionsOrchardZSA = []
+        if have_orchard_zsa:
+            for _ in range(rand.u8() % 5):
+                self.vActionsOrchardZSA.append(OrchardZSAActionDescription(rand))
+            self.flagsOrchardZSA = rand.u8() & 7 # Only three flag bits are currently defined.
+            self.flagsOrchardZSA |= 4  # Setting enableZSAs to true for these tests
+            if self.is_coinbase():
+                # set enableSpendsOrchard = 0
+                self.flagsOrchardZSA &= 2
+            self.valueBalanceOrchardZSA = rand.u64() % (MAX_MONEY + 1)
+            self.anchorOrchardZSA = PallasBase(leos2ip(rand.b(32)))
+            self.proofsOrchardZSA = rand.b(rand.u8() + 32) # Proof will always contain at least one element
+            self.bindingSigOrchardZSA = RedPallasSignature(rand)
+
+        else:
+            # If valueBalanceOrchard is not present in the serialized transaction, then
+            # v^balanceOrchard is defined to be 0.
+            self.valueBalanceOrchardZSA = 0
+
+        # OrchardZSA Burn Fields
+        self.vAssetBurnOrchardZSA = []
+        if have_burn:
+            for _ in range(rand.u8() % 5):
+                self.vAssetBurnOrchardZSA.append(AssetBurnDescription(rand))
+
+        # ZSA Issuance Fields
+        self.vIssueActions = []
+        if have_issuance:
+            self.ik = IssuanceKeys(rand.b(32)).ik
+            for _ in range(rand.u8() % 5):
+                self.vIssueActions.append(IssueActionDescription(rand, self.ik))
+            self.issueAuthSig = rand.b(64)
+
+    def version_bytes(self):
+        return NU7_TX_VERSION | (1 << 31)
+
+    def orchard_zsa_transfer_field_bytes(self):
```

Comment:
This looks completely identical to orchard.
move to be a function on `base`, potentially parametrized. (i suspect that if the var names are changed as suggested, there will be no need for parameters.

Reply:
the last two lines, here and 1 line in orchard should move to up to `def __bytes__(self):`

Reply:
Yes, done.